torch/utils/checkpoint.py

```diff
@@ -66,23 +78,25 @@ def backward(ctx, *args):
         # Stash the surrounding rng state, and mimic the state that was
         # present at this time during forward.  Restore the surrouding state
         # when we're done.
-        rng_devices = [torch.cuda.current_device()] if ctx.had_cuda_in_fwd else []
-        with torch.random.fork_rng(devices=rng_devices, enabled=preserve_rng_state):
-            if preserve_rng_state:
-                torch.set_rng_state(ctx.fwd_cpu_rng_state)
+        rng_devices = []
+        if ctx.preserve_rng_state and ctx.had_cuda_in_fwd:
+            rng_devices = ctx.fwd_gpu_devices
```

Comment:
Well, `had_cuda_in_fwd` shouldn't be a prerequisite here, because we still should restore the CPU state.

Reply:
I'm not quite sure what you mean.  `rng_devices` only controls the GPU devices for which the state should be saved and restored by the context manager.  If cuda was not used in forward, `rng_devices` should be `[]` (which is how torch.random.fork_rng expects to be told "you don't need to mess with any GPU RNG states").

The torch.random.fork_rng context manager [always saves and restores the CPU state unconditionally](https://github.com/pytorch/pytorch/blob/master/torch/random.py#L108), even if the `devices` kwarg is `[]`.

Reply:
Ohh I see, makes sense. Thanks!